### DISCUSSION TITLE
Specifying test .py files relative to pytest.ini file.

### DISCUSSION THREAD

**Discussion by julian-smith-artifex-com** [pytest-dev/pytest#12993] [Q&A]
I'm using `norecursedirs` to exclude various subdirectories when pytest searches for test files.

However there may be new subdirectories to exclude in future, so i'd prefer to anchor `python_files` at the root of my project (or at the directory containing pytest.ini) so that it will always work without me having to add new exclusions.

Is this possible?

For example a syntax like `python_files = ^tests/test_*.py` would be really helpful.

Thanks for any suggestions.

[BTW i've tried `testspaths` and `testpaths`, but they don't seem to avoid the need for `norecursedirs`.

**Comment by portyu9** [comment]
Hi @julian-smith-artifex-com 

Use `testpaths` to specify the directory and `python_files` to specify the filename pattern:

```ini
[pytest]
testpaths = tests
python_files = test_*.py
```

`python_files` is a **glob-style filename pattern**, not a regular expression and not a path expression.

So this:

```ini
python_files = ^tests/test_*.py
```

doesn't mean "anchor this pattern to the `tests` directory". The `^` has no regex-anchor meaning there.

The responsibilities are separate:

```ini
testpaths = tests
```

tells pytest where to start its default collection when you run:

```bash
pytest
```

and:

```ini
python_files = test_*.py
```

tells it which Python filenames inside the collection tree should be considered test modules.

You can specify multiple directories too:

```ini
[pytest]
testpaths =
    tests
    integration_tests

python_files = test_*.py
```

One important limitation: `testpaths` controls pytest's **default** collection roots. If somebody explicitly passes another path:

```bash
pytest some_other_directory/
```

pytest can collect from the explicitly requested location instead.

If you need to prevent recursion into particular directories as well, use `norecursedirs`.

Also note the option is named:

```text
testpaths
```

not `testspaths`.

Please mark this answer as accepted if it helped, thank you.
